builder/hcloud/step_create_server.go

```diff
@@ -118,26 +116,39 @@ func (s *stepCreateServer) Run(ctx context.Context, state multistep.StateBag) mu
 	if err != nil {
 		return errorHandler(state, ui, "Could not create server", err)
 	}
-	state.Put(StateServerIP, serverCreateResult.Server.PublicNet.IPv4.IP.String())
+
 	// We use this in cleanup
 	s.serverId = serverCreateResult.Server.ID
 
-	// Store the server id for later
-	state.Put(StateServerID, serverCreateResult.Server.ID)
-	// instance_id is the generic term used so that users can have access to the
-	// instance id inside of the provisioners, used in step_provision.
-	state.Put(StateInstanceID, serverCreateResult.Server.ID)
-
 	if err := client.Action.WaitFor(ctx, serverCreateResult.Action); err != nil {
 		return errorHandler(state, ui, "Could not create server", err)
 	}
 	if err := client.Action.WaitFor(ctx, serverCreateResult.NextActions...); err != nil {
 		return errorHandler(state, ui, "Could not create server", err)
 	}
 
+	// Store server data for later
+	server := serverCreateResult.Server
+
+	state.Put(StateServerID, server.ID)
+	// instance_id is the generic term used so that users can have access to the
+	// instance id inside of the provisioners, used in step_provision.
+	state.Put(StateInstanceID, server.ID)
+
+	switch {
+	case !server.PublicNet.IPv4.IsUnspecified():
+		state.Put(StateServerIP, server.PublicNet.IPv4.IP.String())
+	case !server.PublicNet.IPv6.IsUnspecified():
+		state.Put(StateServerIP, server.PublicNet.IPv6.IP.String())
+	case len(server.PrivateNet) > 0:
+		state.Put(StateServerIP, server.PrivateNet[0].IP.String())
+	default:
+		return errorHandler(state, ui, "", fmt.Errorf("Could not find server ip"))
+	}
```

Comment:
Could you document this behaviour for users?

Alternative to the heuristic: In the ansible inventory we have an option [`connect_with`](https://github.com/ansible-collections/hetzner.hcloud/blob/e34ff6d295f5900e5f59c77c2eba0b57266c8682/plugins/inventory/hcloud.py#L52-L65) for the user to decide how they want to connect.